File: 01_query_api/02_example.py
# 02_example_python.py
# Simple example of making an API request in Python
# Pairs with 02_example.R
# Tim Fraser

# This script shows how to:
# - Load an API key from a .env file
# - Make a GET request to an example API
# - Inspect the HTTP status code and JSON response

# 0. Setup #################################

## 0.1 Load Packages ############################

# !pip install requests python-dotenv  # run this once in your environment

import os  # for reading environment variables
import requests  # for making HTTP requests
from dotenv import load_dotenv  # for loading variables from .env

## 0.2 Load Environment Variables ################

# Load environment variables from the .env file in the project root
# This matches the behavior of readRenviron(".env") in 02_example.R
load_dotenv(".env")

# Get the API key from the environment
TEST_API_KEY = os.getenv("TEST_API_KEY")

## 1. Make API Request ###########################

# Execute query and save response as object
response = requests.get(
    "https://reqres.in/api/users/2",
    headers={"x-api-key": TEST_API_KEY},
)

## 2. Inspect Response ###########################

# View response status code (200 = success)
print(response.status_code)

# Extract the response as JSON and print
print(response.json())


# Clear environment (optional in short scripts, but shown for parity
# with the R example that clears its workspace)
globals().clear()

File: 01_query_api/02_example.R
# 02_example_r.R

# This script can be run in R to demonstrate 
# how to make an API request

# Install if you haven’t yet
# install.packages(c(“httr2”, “jsonlite”)) 

# Execute query and save response as object
library(httr2)
library(jsonlite)

# Load environment variables from .env file
readRenviron(".env")
TEST_API_KEY = Sys.getenv("TEST_API_KEY")

# Create request object
req = request("https://reqres.in/api/users/2") |>
  req_headers(`x-api-key` = TEST_API_KEY) |>
  req_method("GET")

# Execute request and store result as object
resp = req_perform(req)

# Check status
resp$status_code # 200 = success
# Return response as a json
resp_body_json(resp)

# Or use this if you want to convert the response to a string first and then to a list
fromJSON(resp_body_string(resp))

# Clear environment
rm(list = ls())

# Exit
# q(save = "no")